contracts/marketpools.js

```diff
@@ -0,0 +1,350 @@
+  if (!api.assert(api.BigNumber(amountIn).gt(0), 'insufficient output amount')
+    || !api.assert(api.BigNumber(liquidityIn).gt(0)
+      && api.BigNumber(liquidityOut).gt(0), 'insufficient liquidity')) return false;
+  const num = api.BigNumber(amountIn).times(liquidityOut);
+  const den = api.BigNumber(liquidityIn).plus(amountIn);
+  const amountOut = num.dividedBy(den);
+  if (!api.assert(api.BigNumber(amountOut).lt(liquidityOut), 'insufficient liquidity')) return false;
+  return amountOut;
+}
+
+async function validateOracle(pool, newPrice, maxDeviation = api.BigNumber('0.01')) {
+  const [baseSymbol, quoteSymbol] = pool.tokenPair.split(':');
+  // eslint-disable-next-line no-template-curly-in-string
+  const baseMetrics = baseSymbol !== "'${CONSTANTS.HIVE_PEGGED_SYMBOL}$'"
+    ? await api.db.findOneInTable('market', 'metrics', { symbol: baseSymbol })
+    : { lastPrice: 1 };
+  // eslint-disable-next-line no-template-curly-in-string
+  const quoteMetrics = quoteSymbol !== "'${CONSTANTS.HIVE_PEGGED_SYMBOL}$'"
+    ? await api.db.findOneInTable('market', 'metrics', { symbol: quoteSymbol })
+    : { lastPrice: 1 };
+  if (!baseMetrics || !quoteMetrics) return null; // no oracle available
+  const oracle = api.BigNumber(baseMetrics.lastPrice).dividedBy(quoteMetrics.lastPrice);
+  const dev = api.BigNumber(newPrice).minus(oracle).abs().dividedBy(oracle);
+  // api.debug(`${oracle} -> ${dev} / ${maxDeviation}`);
+  if (!api.assert(api.BigNumber(dev).lte(maxDeviation), 'exceeded max deviation from order book')) return false;
+  return true;
+}
+
+function validateLiquiditySwap(pool, baseDelta, quoteDelta) {
+  const p = api.BigNumber(pool.quoteQuantity).dividedBy(pool.baseQuantity).toFixed(pool.precision, api.BigNumber.ROUND_DOWN);
+  // api.debug(`P - ${p}`);
+  if (!api.assert(api.BigNumber(api.BigNumber(quoteDelta).dividedBy(baseDelta).toFixed(pool.precision, api.BigNumber.ROUND_DOWN)).eq(p),
+    `constant price ${api.BigNumber(quoteDelta).dividedBy(baseDelta)}, expected ${p}`)) return false;
+  return true;
+}
+
+function validateSwap(pool, baseDelta, quoteDelta, maxSlippage) {
+  const k = api.BigNumber(pool.baseQuantity).times(pool.quoteQuantity).toFixed(pool.precision, api.BigNumber.ROUND_HALF_UP);
+  const baseAdjusted = api.BigNumber(pool.baseQuantity).plus(baseDelta);
+  const quoteAdjusted = api.BigNumber(pool.quoteQuantity).plus(quoteDelta);
+  const p = api.BigNumber(pool.quoteQuantity).dividedBy(pool.baseQuantity).toFixed(pool.precision, api.BigNumber.ROUND_DOWN);
+  const pAdjusted = api.BigNumber(quoteAdjusted).dividedBy(baseAdjusted).toFixed(pool.precision, api.BigNumber.ROUND_DOWN);
+  const slippage = api.BigNumber(pAdjusted).minus(p).abs().dividedBy(p);
+  if (!api.assert(api.BigNumber(slippage).lte(maxSlippage), 'exceeded max slippage for swap')) return false;
+  if (!api.assert(api.BigNumber(api.BigNumber(baseAdjusted).times(quoteAdjusted).toFixed(pool.precision, api.BigNumber.ROUND_HALF_UP)).eq(k),
```

Comment:
To make this bullet-proof, I would give it lee-way, e.g.  abs( k.minus( computed ) ) <= 10^{-precision}

Reply:
oh. Rounding actually is probably equivalent, come to think of it.